### PR TITLE
Add a `with_strategy` tactic

### PR DESCRIPTION
**Pull request by JasonGross**
Useful for guarding calls to `unfold` or `cbv` to ensure that, e.g., `Opaque foo` doesn't break some automation which tries to unfold `foo`.

This is a version of #11029 that doesn't persist changes into the global environment and is a tactical rather than a tactic.  However, in order to play nicely with `abstract`, we still need to modify the global environment.

**Kind:** feature

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

Assigning this to @ppedrot because he was the assignee of #11029

This should satisfy @MSoegtropIMC 's desires for unfolding, even if it doesn't satisfy the use-case of my rewriter / fiat-crypto.

### CODE REVIEW

File: plugins/ltac/extratactics.mlg
Comment (ejgallego) [inline plugins/ltac/extratactics.mlg:1149]: I'm very sorry but I'm working hard on removing access to `Global` from tactics so I'd rather no see new calls added.

**Review by ejgallego — DISMISSED** [review]

File: plugins/ltac/extratactics.mlg
Comment (JasonGross) [reply plugins/ltac/extratactics.mlg:1149]: When the call in `abstract` is removed, the call in this tactic can be removed at the same time.  The effect should not be noticeable except via the `abstract` tactic (and other tactics that inspect the global env); I added the calls to the global env just to support `abstract` (and kept them in a separate commit, so it should be easy to revert the changes that commit makes to this file). (And note that the code to set the proof-local env is already in place, so you won't even need to replace the global calls with anything once `abstract` no longer relies on the global env.)  I can add a comment to this effect.  Would this be sufficient?  Or am I missing something about why you want to not see new calls added?

The alternative would be to drop the second commit from this PR (but then we'll have to document that the interaction with `abstract` is broken, or else I can just open an issue about it).

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg:1149]: Thanks for the answer @JasonGross , let me think about it.

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg:1149]: Overall I'd say that either adding the comment or leaving abstract broken should work; my fear with the first option that we suddenly get other parts of the system to depend on the global state so migration to a purely functional tactic state becomes harder.

Having tactics to behave purely functional is IMO extremely important as that will allow Coq to profit from the upcoming OCaml multicore and that is feature we _all_ want to have ASAP :) but indeed, there are tradeoffs.

Comment (JasonGross) [reply plugins/ltac/extratactics.mlg:1153]: I've updated the code and added a bunch of `TODO`s with instructions on what to remove once `abstract` no longer depends on `Global`.

Indeed, I support moving to a purely functional tactic state (though I do think there should be ways to update the document state from tactics, such as `abstract` and #11029, but it seems better to implement this in a functional way).

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg:1153]: Your point about abstract is an interesting one; cc #9146 ; I am not sure the view where abstract does modify document state is the one we want; I'd rather see core calculus let extended such that:
- `let x := foo in y` does allow to specify a transparency in the term; this is enforced for everything.
- `let x := foo in y` does have a trusted mode where `foo` is only type-checked once.

I fact I did a prototype of this last one as it was satisfactory, only problem is that it wasn't easy to design an API where this invariant can be enforced.

Comment (JasonGross) [reply plugins/ltac/extratactics.mlg:1153]: That sounds like it is in the right direction, and I have thoughts about it, but I'll hold them for a venue that's actually about `abstract` rather than sidetracking this PR more.

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg:1153]: It would great to hear your thoughts on abstract tho, a rework of how it works is likely to happen soon so any input is very welcome.

Regarding this PR, my preference sits in the option of "leaving abstract broken" as I'm unsure how this would play with future plans otherwise, but I'll let @ppedrot to have the final say on this, thus his review should override mine.

Comment (ppedrot) [reply plugins/ltac/extratactics.mlg:1153]: > Having tactics to behave purely functional is IMO extremely important as that will allow Coq to profit from the upcoming OCaml multicore

... except that tactics are the least likely part of the system to take advantage of this feature. By definition they are a state monad where the state is a heap of blobs, so the kind of stuff really hard to run in parallel. I am much more confident that we can put multicore to good use in the kernel, which happens to be already purely functional (mostly).

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg:1153]: Not at all @ppedrot ; once tactics are functional you can hope to run the bits between `Proof.` and `Qed.` in a separate thread; with some other minor tricks you can indeed hope to _build_ opaque proofs in their own thread _à la_ isabelle.

Comment (ppedrot) [reply plugins/ltac/extratactics.mlg:1153]: Ah, indeed. I thought you wanted to have fine-grained parallelism in tactics à la `par:`.

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg:1153]: Oh good point, yesterday I was thinking about `par:` as is one of the few bits handled in code I'm replacing that I was not sure how to handle; not sure what should be done here.

**Comment by JasonGross** [comment]
The CI now passes.  What remains to be done to get this merged?

**Comment by MSoegtropIMC** [comment]
I am very much looking forward to this, but I must admit I don't understand the documentation. I would use this for controlling reductions like cbv or cbn e.g. such that certain abstraction borders are maintained - say custom simplification functions. I don't see why I should abstract such terms then. I guess to understand the documentation one has to know some details about efficiency or other issues which happen during Qed or Defined. Maybe an example would help to understand what you mean with your notes in the documentation.

**Comment by JasonGross** [comment]
Suppose you have `Strategy 10 [id]` outside your proof, and you define the factorial function `fact : nat -> nat`.  If you do `assert (id (fact 100) = fact 100) by reflexivity`, this will take about 100!/10! seconds (because doing this with `fact 10` takes a couple seconds and the time complexity is n!) because it will fully reduce `fact` before unfolding `id`.  Note that behavior will be the same if you mark `id` as `Opaque` because conversion treats `Opaque` as merely a hint to unfold this constant last.  If you instead do `assert (id (fact 100) = fact 100) by with_strategy 10 [id] reflexivity`, it will succeed instantly, but the final `Qed` or `Defined` will take 100!/10! seconds, because at the time you hit the end of the proof `id` is marked as `Opaque`.  But if you do `assert (id (fact 100) = fact 100) by with_strategy 10 [id] abstract reflexivity`, *and* you end your proof with `Defined` (needed because ending the proof with `Qed` causes all abstracted subproofs to be inlined and rechecked), then the `assert` will be quick, and the `Defined` will also be quick because there is no rechecking of convertibility.  On small examples this sort of behavior doesn't matter, but because Coq is a super-linear performance domain in so many places, unless you're careful, your may find that your reduction-using tactics are not robust to scaling.

Is this explanation and example, or some subset of it, worth including in the documentation?

**Comment by MSoegtropIMC** [comment]
> Is this explanation and example, or some subset of it, worth including in the documentation?

Yes this is excellent, thanks! Please include it in the documentation.

**Comment by JasonGross** [comment]
Okay, I've worked the example into the note, which is a significant elaboration.  What do you think?  And how do I look at the generated documentation @coq/doc-maintainers ?

**Comment by Zimmi48** [comment]
> Okay, I've worked the example into the note, which is a significant elaboration. What do you think? And how do I look at the generated documentation @coq/doc-maintainers ?

The build for the documentation failed, so that's why there is no link to the output in the status checks section. The problem is apparently your `Fail Timeout 1 Defined.` which results in "The command has not failed." Generally speaking assuming that a timeout is going to be reached or not seems like a bad idea to me.

File: doc/sphinx/proof-engine/ltac.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/ltac.rst:1106]:
```suggestion
      with :cmd:`Qed` rather than :cmd:`Defined`.
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst:3422]: If you could replace `@tactic` with the correct `@ltac_expr`, it would be more precise.

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst:3437]:
```suggestion
      `#12179 <https://github.com/coq/coq/issues/12179>`_.
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst:3483]:
```suggestion
      :g:`*` and :g:`+` before it unfolds :g:`id`, resulting in a full
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst:3525]:
```suggestion
      :cmd:`Defined` for this to work; see the warnings in the
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst:3526]:
```suggestion
      documentation of the :tacn:`abstract` tactic for more details.
```

**Review by Zimmi48 — COMMENTED** [review]
I am not able to review the content (so to speak) of the documentation and only reviewed mostly the form.

**Comment by JasonGross** [comment]
> The problem is apparently your `Fail Timeout 1 Defined.` which results in "The command has not failed." Generally speaking assuming that a timeout is going to be reached or not seems like a bad idea to me.

@Zimmi48 I agree in general, but I think this case is an exception.  My curve-fitting suggests that the `Defined` should take about 10<sup>141</sup> *millennia* (this is about 10<sup>134</sup> times the age of the universe).  So if it's not timing out, that means I set up the example wrong.

**Comment by JasonGross** [comment]
> The problem is apparently your `Fail Timeout 1 Defined.` which results in "The command has not failed." Generally speaking assuming that a timeout is going to be reached or not seems like a bad idea to me.

This is a bug in Coq.  Consider this code:
```coq
Fixpoint fact (n : nat) : nat :=
  match n with
  | 0 => 1
  | S n' => n * fact n'
  end.

Strategy 1000 [id].

Goal True.
  Time assert (id (fact 100) = fact 100) by with_strategy -1 [id] reflexivity.
  exact I.
```
If I do `Timeout 1 Defined.`, I get `Error: Timeout!` after 1 second.  If I do `Fail Timeout 1 Defined.`, I immediately get `Error: The command has not failed!`.  If I do `Fail Fail Timeout 1 Defined.`, I get `Unnamed_thm is defined` followed by `The command has indeed failed with message: The command has not failed!`.  I get the same behavior with
```coq
  Time assert (id (fact 100) = fact 100) by exact_no_check (eq_refl (fact 100)).
  with_strategy -1 [id] exact I.
```
If I remove `with_strategy -1 [id]`, then `Fail` starts behaving correctly.  What's going on here?

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst:3423]:
```suggestion
.. tacn:: with_strategy @strategy_level [ {+ @qualid } ] @ltac_expr
```

**Comment by JasonGross** [comment]
cc @coq/vernac-maintainers

**Comment by JasonGross** [comment]
@Zimmi48 The doc now builds, doc pages [here](https://coq.gitlab.io/-/coq/-/jobs/527375641/artifacts/_install_ci/share/doc/coq/sphinx/html/proof-engine/tactics.html#coq:tacn.with-strategy). Comments and suggestions welcome, especially on which things I'm displaying in the coqtop examples. 
image
image
image
image
image
image

Question: Why is it that `Fail unfold id.`
gives
```
The command has indeed failed with message:
Cannot coerce id to an evaluable reference.
```
twice?

**Comment by JasonGross** [comment]
> Question: Why is it that `Fail unfold id.`
> gives
> 
> ```
> The command has indeed failed with message:
> Cannot coerce id to an evaluable reference.
> ```
> 
> twice?

Ah, the answer is that it's because `.. coqtop:: all reset abort ` does not print any input on lines that follow comments, such as
```coq
         (** This should work, but does not *)
         Fail with_strategy transparent [id] unfold id.
```
Why is this @coq/doc-maintainers ?

**Comment by MSoegtropIMC** [comment]
@JasonGross : just wanted to say that the documentation is very good! As I mentioned I would use this for entirely different purposes, e.g. term simplification which brings terms into specific forms at a specific level of abstraction, which generally requires to make definitions at lower levels of abstraction opaque temporarily. You might want to mention this use - your doc almost reads that it would not be useful for that, but I think it would.

One more note: since for my purpose I need to handle large sets of symbols this would be more manageable in Ltac2 where strategies can be computed e.g. from lists of symbols. Do you plan to have this also in Ltac2, so that it takes an Ltac2 strategy object?

**Comment by Zimmi48** [comment]
> > Question: Why is it that `Fail unfold id.`
> > gives
> > 
> > ```
> > The command has indeed failed with message:
> > Cannot coerce id to an evaluable reference.
> > ```
> > 
> > twice?
> 
> Ah, the answer is that it's because `.. coqtop:: all reset abort ` does not print any input on lines that follow comments, such as
> ```coq
>          (** This should work, but does not *)
>          Fail with_strategy transparent [id] unfold id.
> ```
> Why is this @coq/doc-maintainers ?

That's a bug. Can you open an issue about it?

**Comment by JasonGross** [comment]
@MSoegtropIMC I've added the sentence "This can be useful for guarding calls to reduction in tactic automation to ensure that certain constants are never unfolded by tactics like :tacn:`simpl` and :tacn:`cbn` or to ensure that unfolding does not fail."  The bulk of the warning here is about performance issues that arise when scaling automation.

> since for my purpose I need to handle large sets of symbols this would be more manageable in Ltac2 where strategies can be computed e.g. from lists of symbols. Do you plan to have this also in Ltac2, so that it takes an Ltac2 strategy object?

I don't have plans to implement this, but I strongly support it.  Perhaps you can move `with_set_strategy` to another file as `tclWITH_STRATEGY` and wire that in Ltac2.  I wonder, though, if it might make more sense to just expose `tclENV`, `tclSETENV`, `Environ.oracle`, `Environ.set_oracle`, `Conv_oracle.get_strategy`, and `Conv_oracle.set_strategy` in Ltac2 (which after all is supposed to be a much more low-level scripting language, I think), and then implement `with_strategy` in Ltac2 user-land.  What do you think @MSoegtropIMC @ppedrot ?

**Comment by JasonGross** [comment]
I've discovered another bug in the implementation; the following example does not work as expected:
```coq
(* test that the strategy is correctly reverted after closing the goal completely *)
Goal id 0 = 0.
  assert (id 0 = 0) by with_strategy expand [id] reflexivity.
  Fail unfold id. (* has not failed *)
  reflexivity.
Qed.

(* test that the strategy is correctly reverted after failure *)
Goal id 0 = 0.
  let id' := id in
  (try with_strategy expand [id] fail); assert_fails unfold id'.
  Fail unfold id. (* has not failed *)
  reflexivity.
Qed.
```

Presumably this is because
```ocaml
  Proofview.Goal.enter begin
    fun _ ->
      List.iter (fun (k, lvl) -> Global.set_strategy k lvl) orig_kl_global;
      Proofview.tclUNIT ()
  end
```
does not work when there are no goals, and because
```
  Proofview.Goal.enter begin
    fun _ ->
      List.iter (fun q -> Global.set_strategy (glob_key q) lvl) ql;
      Proofview.Unsafe.tclSETENV env
  end
```
is not properly captured by the ProofView monad in a way that plays nicely with backtracking.  (I bet this also means that `with_strategy` doesn't play nicely with muti-success tactics and backtracking points inside its execution...  Thoughts @ejgallego ?

**Comment by JasonGross** [comment]
Here's a test for multi-success, by the way (which fails as expected) :
```coq
(* test multi-success *)
Goal id (fact 100) = fact 100.
  Timeout 1
          (with_strategy -1 [id] (((idtac + (abstract reflexivity))); fail)).
  Undo.
  Timeout 1
          (with_strategy -1 [id] (idtac + (abstract reflexivity))); fail. (* should not time out *)
Qed.
```

**Comment by JasonGross** [comment]
I have fixed all three of these bugs and added test-cases for them.  This required exposing two new primitives in Proofview (ones which I'm not sure how to implement without access to `split`, which seems to only be accessible in a well-typed way from inside Proofview): `tclFINALLY` (which is just a slightly improved version of a tactical previously implemented inside the Ltac profiler) and a somewhat clunky tactical I've called `tclBEFOREEACHSUCCESS`.  Hopefully this is now ready for review and merging.

**Comment by MSoegtropIMC** [comment]
> What do you think @MSoegtropIMC

I must admit I don't have time right now to look into the details, but I for sure will.

**Comment by Zimmi48** [comment]
@JasonGross There's now some conflict to fix in the syntax extensions chapter (after merging #11718).

**Comment by JasonGross** [comment]
I've fixed the syntax conflicts, but left a `todo pr` about where to document `strategy_level_or_var`.  I've also found a new proofview primitive that I'm much happier with, which I'm tentatively calling `tclWRAPFINALLY`.  Happily, it allows solving both the issues here as well as the issues with multi-success backtracking in LtacProf; I've made #12197 for that, and both of these PRs share the commit that adds this primitive to Proofview.

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/vernacular-commands.rst:960]: The location looks good to me. You might want to try to regenerate the grammar files `fullGrammar` and `orderedGrammar` following your changes. Then you might need to manually move the rules for `strategy_level_or_var` next to `strategy_level` (in `orderedGrammar`). Have a look at `doc/tools/doc_grammar/README.md`. No worries if you don't have time to do this (this is not checked by the CI and can be adjusted by Jim or I later on).

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1714]: Was it already supported and not documented or is this added by this PR? If that's the latter, then you might want to consider renaming it `smart_qualid`: farther from the implementation but closer to the documentation.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1787]:
```suggestion
             is the reason for introducing special entries
```

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [reply doc/sphinx/user-extensions/syntax-extensions.rst:1714]: Introduced by this PR.  We currently call `qualid` a `reference` in tactic-notation-land, though...  I feel like `smart_global` is less opaque than `smart_qualid`.  Are we planning to replace `reference` with `qualid` someday?  In any case, I'm happy to bow to consensus on this.

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/vernacular-commands.rst:960]: I will leave this to you and Jim; when I run `make doc_gram; make doc_gram_rsts`, it *removes* `strategy_level_or_var` from the .rst file, adds a `| @strategy_level0` entry to `strategy_level` (which exists in the code to "work around a limitation of the macro system" (cc @ppedrot )), and I get a bunch of warnings:
```
DOC_GRAM_RSTS
# cmds in rsts, gram, total = 275 243 422
# tacs in rsts, gram, total = 267 0 267
Warning: editedGrammar: Undefined symbol 'strategy_level0'
Warning: editedGrammar: Unreachable symbol 'dangling_pattern_extension_rule'
Warning: editedGrammar: Unreachable symbol 'vernac_aux'
Warning: editedGrammar: Unreachable symbol 'subprf'
Warning: editedGrammar: Unreachable symbol 'finite_token'
Warning: editedGrammar: Unreachable symbol 'tactic_mode'
Warning: editedGrammar: Unreachable symbol 'ltac_use_default'
Warning: editedGrammar: Unreachable symbol 'tacticals'
Warning: Singleton non-terminal, maybe SPLICE?: first_letter
Warning: Singleton non-terminal, maybe SPLICE?: ltac_expr_opt
Warning: Singleton non-terminal, maybe SPLICE?: num
Warning: Singleton non-terminal, maybe SPLICE?: scope_key
Warning: Singleton non-terminal, maybe SPLICE?: scope_name
Warning: Singleton non-terminal, maybe SPLICE?: setting_name
Warning: Singleton non-terminal, maybe SPLICE?: subsequent_letter
Warning: Singleton non-terminal, maybe SPLICE?: type
Warning: Singleton non-terminal, maybe SPLICE?: vernacular
Warning: doc/sphinx/proof-engine/tactics.rst line 3950: Repeated cmd: 'Print Hint @ident'
Warning: Nonterminals 'tactic_mode' to 'ltac_use_default' not included in .rst files (2)
Warning: Nonterminal dangling_pattern_extension_rule not included in .rst files
Warning: Nonterminal type not included in .rst files
Warning: Nonterminals 'firstorder_rhs' to 'where' not included in .rst files (2)
Warning: Nonterminals 'control_command' to 'tacticals' not included in .rst files (3)
Warning: Nonterminals 'vernac_aux' to 'subprf' not included in .rst files (2)
Warning: Nonterminal finite_token not included in .rst files
Warning: Nonterminals 'scheme_kind' to 'hint_info' not included in .rst files (3)
Warning: Nonterminal strategy_level_or_var not included in .rst files
Warning: Nonterminals 'command' to 'starredidentref' not included in .rst files (6)
Warning: Nonterminal comment not included in .rst files
Warning: Nonterminals 'hint' to 'tacdef_body' not included in .rst files (2)
Warning: Nonterminals 'int_or_id' to 'fun_scheme_arg' not included in .rst files (3)
Warning: Nonterminal hints_path not included in .rst files
Warning: Nonterminals 'simple_tactic' to 'occurrences' not included in .rst files (44)
Warning: Nonterminals 'ltac_expr' to 'match_hyp' not included in .rst files (28)
TOTAL Nonterminals not included = 101
```

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [reply doc/sphinx/user-extensions/syntax-extensions.rst:1714]: > Are we planning to replace `reference` with `qualid` someday?

That would be the idea (it is already like this everywhere in the doc), but calling @herbelin to weigh on this.

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (Zimmi48) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: OK, will take care of this, possibly by pushing to this very PR if you'd like.

Comment (JasonGross) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: Yes, please, and thanks!

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [reply doc/sphinx/user-extensions/syntax-extensions.rst:1714]: Hmm, to do this, do I change
```ocaml
let () = create_generic_quotation "smart_global" Pcoq.Prim.smart_global Stdarg.wit_smart_global
```
to
```ocaml
let () = create_generic_quotation "smart_qualid" Pcoq.Prim.smart_global Stdarg.wit_smart_global
```
or do I need to do more than this?

**Comment by JasonGross** [comment]
So, uh, it seems that the test-suite reliably times out on MacOS?!  I don't get it, though, the failing build log ends with
```
TEST      modules/sub_objects.v 
CHECK     modules/sub_objects.v
TEST      modules/subtyping.v 
CHECK     modules/subtyping.v
##[error]The operation was canceled.
Finishing: Run Coq Test Suite
```
(the stage takes nearly 44m)

whereas a recent success completed the test-suite in under 10m and the log ends with
```
TEST      modules/sub_objects.v 
CHECK     modules/sig.v
TEST      modules/subtyping.v 
CHECK     modules/sub_objects.v
CHECK     modules/subtyping.v
Makefile:92: cannot run complexity tests (no bogomips found)
BUILDING SUMMARY FILE
Makefile:92: cannot run complexity tests (no bogomips found)
NO FAILURES

Finishing: Run Coq Test Suite
```
Can someone with a MacOS machine confirm?

**Comment by JasonGross** [comment]
Oh, wait, this is not just a MacOS issue, also happens on Linux :-(

**Comment by JasonGross** [comment]
Oh, I see the issue.  The strategy directives are not recorded in the .vo file, resulting in the test-suite file timing out on coqchk.

**Comment by JasonGross** [comment]
Okay, I've fixed the issue.  I actually have two different fixes.  First I added a commit that made the tactic add info to the libobject stack (I'm not actually sure this fixed the issue, though; I seem to recall it fixing the issue, but thinking back on it, it seems like it shouldn't).  Then, because @ppedrot told me before at https://github.com/coq/coq/pull/11029#discussion_r355141862 that no tactic should mess with the libobject stack, I added a revert of that commit and instead just disabled coqchk on the relevant test-suite file, and added a warning to the manual.  (In the process, I found a new bug in Coq, which I reported as #12200.)

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst:1714]: #11840 was going in the direction of `smart_global`. The point is also that the section on tactic notations is as well on syntax as on semantics while the name `qualid` is mostly a syntax thing.

Additionally, a string is not technically a qualid (even though we could eventually think at allowing qualified strings such as `File.Mod."+"`).

Comment (Zimmi48) [reply doc/sphinx/user-extensions/syntax-extensions.rst:1714]: OK, then I guess we'll backtrack on the renaming in the doc. What about `smart_ref` / `smart_reference`? It's not clear immediately what `global` stands for (I guess it means "global reference") and I don't think it is used much in any user-facing doc / features so far. BTW, it looks like `ref` is supported there, not just `reference`. Do they mean the same thing?

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (Zimmi48) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: Done.

File: interp/stdarg.ml
Comment (ppedrot) [inline interp/stdarg.ml:45]: You need to share the same runtime representation as `wit_ref` using the `~dyn` optional argument.

Comment (JasonGross) [reply interp/stdarg.ml:45]: I'm not sure what you mean, can you make an edit suggestion?

Comment (ppedrot) [inline interp/stdarg.ml:45]:
```suggestion
let wit_smart_global = make0 ~dyn:(val_tag (topwit wit_ref)) "smart_global"
```

Comment (JasonGross) [reply interp/stdarg.ml:45]: Happy to commit this (thanks), but what does this do?  (Or where's the doc?)

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: Thanks!

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (JasonGross) [reply doc/sphinx/user-extensions/syntax-extensions.rst:1714]: I'm happy to rename `smart_global` to `smart_reference`, or to make `reference` be smart and introduce `qualid` for what is currently `reference`, and change most things over to be (smart) `reference`. Would it be better to converge on #11840 first?  What's the status of #11840?

(And, yes, `ref` and `reference` are the same, I believe, see also #9122

File: doc/sphinx/proof-engine/tactics.rst
Comment (jfehrle) [inline doc/sphinx/proof-engine/tactics.rst:3363]: How about a more concrete statement along these lines (if correct)?:

```
Executes :token:`ltac_expr3`, applying the alternate unfolding behavior
that :cmd:`Strategy` command controls, but only for :token:`ltac_expr3`.
This can be useful ...
```

**Review by jfehrle — COMMENTED** [review]
One wording suggestion.

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst]: Sure, thanks, and updated

**Comment by JasonGross** [comment]
I've just rebased this on top of master and fixed some merge conflicts.  What remains to be reviewed?  I'd like to get this in 8.12+beta1.  (@ppedrot , do you think this is doable?)

**Review by ppedrot — COMMENTED** [review]
I don't like the fact that a tactic implementation lives in the mlg file. I know we do have a few of those here and there, but it's mostly an artifact due to historical reasons. In particular, the code of this tactic is fairly big. You should probably move it somewhere else. Is it technically possible to have it into `tactics.ml` for instance?

**Comment by ppedrot** [comment]
This is mostly ready except for my above nit. In any case, this should definitely be part of 8.12.

**Comment by JasonGross** [comment]
I've moved it to tactics.ml (and reworked the history for it to have started there, since I think the evolution of this tactic may be informative in the future).  I've also left a note above `tclWRAPFINALLY` (which now also lives in `tactics.ml`, but is not exported in the `.mli`) about the confusion of where it should live.  Hopefully this is ready now.

**Review by ppedrot — APPROVED** [review]

**Comment by JasonGross** [comment]
The Mac job is failing with
```
File "./modules/PO.v", line 51, characters 0-12:
Error:
/Users/runner/runners/2.166.4/work/1/s/test-suite/modules/Nat.vo: premature end of file. Try to rebuild it.

0m0.001s 0m0.001s
0m0.084s 0m0.050s
==========> FAILURE <==========
    modules/PO.v...Error! (should be accepted)

FAILURES
    modules/PO.v...Error! (should be accepted)
```
which seems unrelated.  The rest of the CI passes, so this should be ready for merge?

**Comment by ppedrot** [comment]
Let's merge.

**Pull request merged**
